shotgun_api3/shotgun.py

```diff
@@ -3539,6 +3526,137 @@ def _dict_to_extra_data(self, d, key_name="value"):
         """
         return dict([(k, {key_name: v}) for (k,v) in (d or {}).iteritems()])
 
+    def _upload_file_to_storage(self, path, storage_url):
+        """
+        Internal function to upload an entire file to the Cloud storage.
+        
+        :param str path: Full path to an existing non-empty file on disk to upload.
+        :param str storage_url: Target URL for the uploaded file. 
+        """
+        filename = os.path.basename(path)
+
+        fd = open(path, "rb")
+        try:
+            content_type = mimetypes.guess_type(filename)[0]
+            content_type = content_type or "application/octet-stream"
+            file_size = os.fstat(fd.fileno())[stat.ST_SIZE]
+            self._upload_data_to_storage(fd, content_type, file_size, storage_url )
+        finally:
+            fd.close()
+
+        LOG.debug("File uploaded to Cloud storage: %s", filename)
+
+    def _multipart_upload_file_to_storage(self, path, upload_info):
+        """
+        Internal function to upload a file to the Cloud storage in multiple parts.
+
+        :param str path: Full path to an existing non-empty file on disk to upload.
+        :param dict upload_info: Contains details received from the server, about the upload.
+        """
+
+        fd = open(path, "rb")
+        try:
+            content_type = mimetypes.guess_type(path)[0]
+            content_type = content_type or "application/octet-stream"
+            file_size = os.fstat(fd.fileno())[stat.ST_SIZE]
+            filename = os.path.basename(path)
+
+            etags = []
+            part_number = 1
+            bytes_read = 0
+            chunk_size = 20000000
+            while bytes_read < file_size:
+                data = fd.read(chunk_size)
+                bytes_read += len(data)
+                part_url = self._get_upload_part_link(upload_info, filename, part_number)
+                etags.append(self._upload_data_to_storage(data, content_type, len(data), part_url ))
+                part_number += 1
+
+            self._complete_multipart_upload(upload_info, filename, etags)
+        finally:
+            fd.close()
+
+        LOG.debug("File uploaded in multiple parts to Cloud storage: %s", path)
+
+    def _get_upload_part_link(self, upload_info, filename, part_number):
+        """
+        Internal function to get the url to upload the next part of a file to the
+        Cloud storage, in a multi-part upload process.
+
+        :param dict upload_info: Contains details received from the server, about the upload.
+        :param str filename: Name of the file for which we want the link.
+        :param int part_number: Part number for the link.
+        :returns: upload url.
+        :rtype: str
+        """
+        params = {
+            "upload_type": upload_info["upload_type"],
+            "filename": filename,
```

Comment:
curious - why is the filename important here? i am guessing that the server relies on the `upload_id` to piece things together... why do we need to send the filename as well?

might be worth adding some comments to explain